### PR TITLE
✨🐛 HC32 - Add SERIAL_DMA, fix SDIO and MEATPACK

### PR DESCRIPTION
**Pull request by shadow578**
### Description

this PR adapts the HC32 HAL to make use of some new features in the arduino core and ddl, including support for `SERIAL_DMA`.
It also introduces some extra checks aimed to improve error messages.

### Requirements

- any supported HC32-based board.
- updating the packages [`framework-arduino-hc32f46x`](https://github.com/shadow578/framework-arduino-hc32f46x/) and [`framework-hc32f46x-ddl`](https://github.com/shadow578/framework-hc32f46x-ddl) to their latest release versions (`1.1.0` and `2.2.1` respectively).
  - using a older version of `framework-arduino-hc32f46x` will cause an error prompting the user to update.

### Benefits

- general improvements in Serial communication\* and support for `SERIAL_DMA` 
- automatic configuration of the arduino core based on Marlins `Configuraton.h` and `Configuration_adv.h`
  - e.g. enabling core debugging when enabling `MARLIN_DEV_MODE` 
- fixed support for meatpack

During my testing (about 10 hours of printing as of now), serial errors when host printing were reduced to 0%.

\* see https://github.com/shadow578/framework-arduino-hc32f46x/pull/10 for details on the 'general improvements' part

### Configurations

[sample configuration for Aquila X2 w/ DWIN MarlinUI + BLTouch + SERIAL_DMA](https://github.com/MarlinFirmware/Marlin/files/14498688/Configuration.zip)

### Related Issues

N/A

### CODE REVIEW

**Comment by classicrocker883** [comment]
unrelated to this **PR** specifically, but HC32 related:  
I have run into an issue after flashing the firmware I would get a continual restart loop while the SD card is inserted. once removed it can be used as normal, otherwise anytime the SD card is put in it reboots over and over. is that still an issue?

**Comment by shadow578** [comment]
> unrelated to this **PR** specifically, but HC32 related: I have run into an issue after flashing the firmware I would get a continual restart loop while the SD card is inserted. once removed it can be used as normal, otherwise anytime the SD card is put in it reboots over and over. is that still an issue?

@classicrocker883  thanks for bringing this issue to my attention. 
after looking into the sdio HAL, i've noticed that with every initialization of the SDIO (= every time a sd card is inserted), the card handle was leaked.
on repeated sd card insertions, this would cause the system to run out of memory and crash.

anyway, i've created a fix and added the commit to this PR (even if it's not related...)

---
EDIT: 

just noticed that @classicrocker883's issue occurs on boot. 
maybe increasing the available heap memory (using [`DDL_HEAP_SIZE`](https://github.com/shadow578/framework-hc32f46x-ddl#custom-stack-and-heap-sizes)) could help?

if not, it'd probably be best to move the discussion to a issue

**Comment by classicrocker883** [comment]
> just noticed that @classicrocker883's issue occurs on boot. maybe increasing the available heap memory (using [`DDL_HEAP_SIZE`](https://github.com/shadow578/framework-hc32f46x-ddl#custom-stack-and-heap-sizes)) could help?
> 
> if not, it'd probably be best to move the discussion to a issue

as for the value to change what do you suggest? in that link there are two different values presented, default and example. im not sure where to start.

**Comment by shadow578** [comment]
> as for the value to change what do you suggest? in that link there are two different values presented, default and example. im not sure where to start.

the default values are what you'd be using right now, so i'd suggest setting them like so:

```ini
[HC32F460_base]
# (...)
build_flags      =
  -D DDL_STACK_SIZE=0x800 -D DDL_HEAP_SIZE=0x2000
  -D ARDUINO_ARCH_HC32
  -D PLATFORM_M997_SUPPORT
# (...)
```

when everything worked, you should see that the amount of free memory that marlin prints on start increased by ~5120 bytes.
so when before you saw "echo: Free Memory: 1800" before the change, after you should see "echo: Free Memory: 6920".

please note that for this to do anything at all, you **must** use version 2.2.1 of the ddl.

**Comment by classicrocker883** [comment]
something isnt working with **app_config.h**, I had to revert **hc32.ini** and things at least compiled normally.

in that file, I see:
```y
#if ENABLED(POSTMORTEM_DEBUGGING)
  // disable arduino core fault handler, as we define our own
  #define CORE_DISABLE_FAULT_HANDLER 1
#endif
```
yet this is automatically enabled/defined in **hc32.ini**
```ini
  -D CORE_DISABLE_FAULT_HANDLER             # Disable arduino core fault handler (we use our own)
```

should `POSTMORTEM_DEBUGGING` also be enabled or can `-D CORE_DISABLE_FAULT_HANDLER` be commented out?

**Comment by shadow578** [comment]
@classicrocker883 
> something isnt working with app_config.h, I had to revert hc32.ini and things at least compiled normally.

make sure you're using the latest release (at least 1.1.0) of the arduino core.

> in that file, I see:
>
> #if ENABLED(POSTMORTEM_DEBUGGING)
>   // disable arduino core fault handler, as we define our own
>  #define CORE_DISABLE_FAULT_HANDLER 1
> #endif
>
> yet this is automatically enabled/defined in hc32.ini
>
>  -D CORE_DISABLE_FAULT_HANDLER             # Disable arduino core fault handler (we use our own)
>
> should POSTMORTEM_DEBUGGING also be enabled or can -D CORE_DISABLE_FAULT_HANDLER be commented out?

this is indeed a change in the default behaviour of the HC32 HAL which i honestly didn't really consider. 
as per the [fault handler documentation](https://github.com/shadow578/framework-arduino-hc32f46x/blob/main/docs/FAULT_HANDLER.md), `CORE_DISABLE_FAULT_HANDLER` will disable the handler built into the arduino core.

so the old behaviour was to just always remove the fault handler, meaning a fault would just lock up the system.
with the change, you'll always get a fault report (at the cost of some flash usage).


imo the new behaviour is a better default, as you'll be made aware of the issue.

**Comment by InsanityAutomation** [comment]
@shadow578 Is this still in a good state in youre opinion? Was looking at getting the Tenlog HC32 board ported in soon and it does seem to call in some of the stuff done here in their fork so figured this would be a prereq.

**Comment by shadow578** [comment]
> @shadow578 Is this still in a good state in youre opinion? Was looking at getting the Tenlog HC32 board ported in soon and it does seem to call in some of the stuff done here in their fork so figured this would be a prereq.

as far as i'm concerned, this should be good to merge and thus in a good state for adapting to a different HC32 board.

but please note that in the end it shouldn't really matter what Tenlog's port "calls in", as the HAL should work for basically every HC32 board with only changes to the config (offset in `hc32.ini` and pins).
i've seen some people trying to port the HAL to different boards and in doing so trying to fork and modify the DDL and stuff. this should __not__ be needed.

**Comment by shadow578** [comment]
@thisiskeithb sorry for asking, but is there anything preventing the PR from getting reviewed / merged?

**Comment by thisiskeithb** [comment]
> sorry for asking, but is there anything preventing the PR from getting reviewed / merged?

We have [nearly 100 open PRs](https://github.com/MarlinFirmware/Marlin/pulls). Please have some patience.

**Comment by InsanityAutomation** [comment]
I'm not a bare metal side expert on the underlying platforms, so I requested Jason review.  I don't see anything that jumps out as wrong to me, but he's the resident expert so I'll leave the review to him

File: ini/hc32.ini [hunk 1]
```diff
@@ -33,22 +33,14 @@ build_src_filter = ${common.default_src_filter} +<src/HAL/HC32> +<src/HAL/shared
...
-  # DDL / Arduino Debug Options
-  #-D __DEBUG                               # DDL debug mode
-  #-D __CORE_DEBUG                          # Arduino core debug mode
-  -D PANIC_ENABLE                           # enable custom panic handlers (in MinSerial)
-  # options to reduce debug mode footprint (-16K; messages are less verbose)
-  -D __DEBUG_SHORT_FILENAMES                # Use short filenames in DDL debug output
-  -D __PANIC_SHORT_FILENAMES                # Use short filenames in core panic output
-  -D __OMIT_PANIC_MESSAGE                   # Omit panic messages in core panic output
+# hc32 app configuration file
+board_build.app_config = Marlin/src/HAL/HC32/app_config.h
```
Comment (sjasonsmith) [inline ini/hc32.ini:43]: I am not familiar with this `board_build.app_config` mechanism, and am having trouble finding documentation on it. Is this unique to the HC32 core, or something that is more broadly available?

If this is something unique to HC32, would you get the same effect by including app_config.h. from `HAL/HC32/HAL.h`?

File: Marlin/src/HAL/HC32/MarlinHAL.cpp [hunk 2]
```diff
@@ -123,6 +123,11 @@ void MarlinHAL::init() {
 
   // Register min serial
   TERN_(POSTMORTEM_DEBUGGING, install_min_serial());
+
+  // warn if low memory after init
+  if (freeMemory() < 1024) {
+    SERIAL_WARN_MSG("HAL: low memory after init!\n");
+  }
```
Comment (sjasonsmith) [inline Marlin/src/HAL/HC32/MarlinHAL.cpp:130]: Did you encounter problems related to this causing you to add this warning message?

Would it be valuable to print the actual number with the warning?

File: Marlin/src/HAL/HC32/MarlinSerial.cpp [hunk 3]
```diff
@@ -46,14 +46,34 @@ constexpr bool serial_handles_emergency(int port) {
...
+// serial port where RX uses DMA and TX uses IRQs
+// all serial ports use DMA1
+// since there are 4 USARTs and 4 DMA channels, we can use the USART number as the DMA channel
+#define DEFINE_DMA_SERIAL_MARLIN(name, n)      \
+  MSerialT name(serial_handles_emergency(n), \
+                &USART##n##_config,          \
+                BOARD_USART##n##_TX_PIN,     \
+                BOARD_USART##n##_RX_PIN,     \
+                M4_DMA1,                     \
+                ((en_dma_channel_t)(n - 1))); // map USART1 to DMA channel 0, USART2 to DMA channel 1, etc.
```
Comment (sjasonsmith) [inline Marlin/src/HAL/HC32/MarlinSerial.cpp:66]: Utilizing DMA in other HALs has been complicated by the wide mixture of boards and configurations, making it extremely hard to avoid resource conflicts in a flexible way.

I know there aren't many boards using the HC32, so that should be a lot less likely.

From your experience, should we be concerned about the longevity of an assumption that DMA channels 0/1 will always be available for the serial ports to use?

File: Marlin/src/HAL/HC32/app_config.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+//
+// Misc.
+//
+
+// redirect printf to host serial
+#define REDIRECT_PRINTF_TO_SERIAL 1
+
+// FIXME override F_CPU to PCLK1, as marlin freaks out otherwise
+#define F_CPU (SYSTEM_CLOCK_FREQUENCIES.pclk1)
```
Comment (sjasonsmith) [inline Marlin/src/HAL/HC32/app_config.h:68]: Can you elaborate on this FIXME comment any more?

Is this a case where F_CPU needs to contain the simple fixed frequency for use in other preprocessor macros, rather than what may be run-time frequency queries normally supported in the framework?

File: Marlin/src/HAL/HC32/sdio.cpp [hunk 5]
```diff
@@ -54,7 +54,7 @@
     fn                                              \
   }
 
-stc_sd_handle_t *handle;
+stc_sd_handle_t *handle = nullptr;
```
Comment (sjasonsmith) [inline Marlin/src/HAL/HC32/sdio.cpp:57]: I would prefer to see the SDIO changes in a separate PR. That way if there are problems with either the SDIO or SERIAL_DMA features they can be readily reverted independent of one another.

I'm not really going to hold the line on these, the changes aren't all that complicated and Marlin has a long history of large commits of unrelated changes.

File: ini/hc32.ini (hunk 1 above)
Comment (shadow578) [reply ini/hc32.ini:43]: `board_build.app_config` is indeed unique to the HC32 arduino core, documentation [see here](https://github.com/shadow578/framework-arduino-hc32f46x/blob/main/docs/APP_CONFIG.md).

it's needed since some parts of the arduino core are configured using defines, and the `app_config.h` is included into all of the core's files

File: Marlin/src/HAL/HC32/MarlinHAL.cpp (hunk 2 above)
Comment (shadow578) [reply Marlin/src/HAL/HC32/MarlinHAL.cpp:130]: i've added this warning in response to one of @classicrocker883's comments in this PR, where he had issues with random crashes (https://github.com/MarlinFirmware/Marlin/pull/26845#issuecomment-1980218771). 

Besides a memory leak in the sdio driver, i've found that depending on the configuraiton the system could run out of the allocated heap space. 
This warning is meant as a small 'heads-up' message, since out-of-memory issues tend to be really hard to debug.

File: Marlin/src/HAL/HC32/MarlinSerial.cpp (hunk 3 above)
Comment (shadow578) [reply Marlin/src/HAL/HC32/MarlinSerial.cpp:66]: from what i can tell, all chips in the HC32 family supported by the HAL (even if some only theoretically) should have two DMA units with 4 channels each. 
So from a hardware side it should be no issue.

From the software side, the only thing that could configure DMA channels beside the Marlin HAL would be a bootloader. 
But that shouldn't be an issue either since 
a) all bootloaders known to me don't use DMA at all, and
b) as part of initialization the DMA peripheral is reset, so any previous configuration is disabled.

File: Marlin/src/HAL/HC32/app_config.h (hunk 4 above)
Comment (shadow578) [reply Marlin/src/HAL/HC32/app_config.h:68]: yes, marlin uses certain timer frequency constants `constexpr` expressions, which requires them to be constant values and why the timer rates in `timers.h` are fixed values (also with TODOs).

for this FIXME, i'm not quite sure what causes the issues, and i honestly cannot remember what exactly the issues were before i've added it, but the printer just seemed to behave weirdly. 

i think it could be that Marlin assumes the CPU frequency and timer frequencies to be equal, which isn't the case for HC32. But i really haven't looked into it too much

File: Marlin/src/HAL/HC32/sdio.cpp (hunk 5 above)
Comment (shadow578) [reply Marlin/src/HAL/HC32/sdio.cpp:57]: ok, i'll keep unrelated changes in separate PRs going forward.

File: Marlin/src/HAL/HC32/app_config.h (hunk 4 above)
Comment (sjasonsmith) [reply Marlin/src/HAL/HC32/app_config.h:68]: I went looking back through past changes I made for STM32 and how these various clocks and variables are used.

I see that you aren't changing F_CPU in this PR, only changing where it is defined, so any clock/timer related changes should happen in another PR. Hopefully these comments are helpful, based on some of the questions I see in the comments in the HC32/timers.h file.

`F_CPU` - This has to be constexpr and is expected to be the rate of the main MCU core, as it is used throughout Marlin in macros such as DELAY_US. Setting this to PCLK1 probably means that anything using those timing macros is going to be running too quicky, unless the peripheral and core clocks are the same.

`timers.h : CYCLES_PER_MICROSECOND` - This is only used for some delays during babystepping, and should be based on the core clock, not the timer clock. I believe this should be deleted from your `timers.h`, allowing the default based on `F_CPU` in `macros.h` to be used.

`timers.h: HAL_TIMER_RATE` - This _does not_ have to be constexpr. It is only used inside the HAL to configure the timers.  I know this because a few years ago I modified the STM32 HAL to query it at runtime, since the clocks are not consistent across timers. I've linked the PR where I originally made that change below:
#19978 : Read STM32 clock rates from framework

I suspect that HAL_TIMER_RATE might have previously been used outside of the HAL, which doesn't seem to be true anymore. I'd consider that macro open to elimination if it no longer serves purpose in a HAL. The only outside usage I see is one STM32 board overriding the value...but I can't imagine that is actually correct.
